algosdk/future/atomic_transaction_composer.py

```diff
@@ -0,0 +1,512 @@
+                will contain None for that method's return value.
+        """
+        if self.status > AtomicTransactionComposerStatus.SUBMITTED:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposerStatus must be submitted or lower to execute a group"
+            )
+
+        self.submit(client)
+        transaction.wait_for_confirmation(client, self.tx_ids[0], wait_rounds)
+        self.status = AtomicTransactionComposerStatus.COMMITTED
+
+        confirmed_round = -1
+        method_results = []
+
+        for i, tx_id in enumerate(self.tx_ids):
+            if i not in self.method_dict:
+                continue
+            # Return is void
+            if self.method_dict[i].returns is None:
+                method_results.append(
+                    ABIResult(
+                        tx_id=tx_id,
+                        raw_value=None,
+                        return_value=None,
+                        decode_error=None,
+                    )
+                )
+                continue
+
+            # Parse log for ABI method return value
+            resp = client.pending_transaction_info(tx_id)
+            confirmed_round = resp["confirmed-round"]
+            logs = resp["logs"] if "logs" in resp else None
+            raw_value = None
+            return_value = None
+            decode_error = None
+            # Look for the last returned value in the log
+            for result in reversed(logs):
+                # Check that the first four bytes is the hash of "return"
+                result_bytes = base64.b64decode(result)
+                if result_bytes[:4] != bytes.fromhex(ABI_RETURN_HASH):
+                    continue
+                try:
+                    raw_value = result_bytes[4:]
+                    return_value = self.method_dict[i].returns.type.decode(
+                        raw_value
+                    )
+                    decode_error = None
+                except Exception as e:
+                    raw_value = None
+                    return_value = None
+                    decode_error = e
+                break
+
+            abi_result = ABIResult(
+                tx_id=tx_id,
+                raw_value=raw_value,
+                return_value=return_value,
+                decode_error=decode_error,
+            )
+            method_results.append(abi_result)
+
+        if confirmed_round < 0:
+            resp = client.pending_transaction_info(self.tx_ids[0])
+            confirmed_round = resp["confirmed-round"]
+
+        return AtomicTransactionResponse(
+            confirmed_round=confirmed_round,
+            tx_ids=self.tx_ids,
+            results=method_results,
+        )
+
+
+class TransactionSigner(ABC):
+    """
+    Represents an object which can sign transactions from an atomic transaction group.
+    """
+
+    def __init__(self) -> None:
+        pass
+
+    def sign(self, txn_group, indexes):
```

Comment:
This should use `@abstractmethod` so all subclasses are forced to override it

Also, how about a name like `sign_transactions` instead? At some point we may want to introduce signers for other objects, and if they all use `sign` it won't be possible for subclasses to implement multiple interfaces.

Reply:
Sounds good!